src/SDDK/wave_functions.hpp

```diff
@@ -329,26 +335,82 @@ class Wave_functions
      *  \param [in] i0   Starting index of wave-functions in src.
      *  \param [in] jspn Spin component on destination wave-functions.
      *  \param [in] j0   Starting index of wave-functions in destination. */
-    void copy_from(device_t pu__, int n__, Wave_functions const& src__, int ispn__, int i0__, int jspn__, int j0__);
+    void copy_from(device_t pu__, int n__, Wave_functions<T> const& src__, int ispn__, int i0__, int jspn__, int j0__);
+
+    template <typename F>
+    void copy_from(device_t pu__, int n__, Wave_functions<F> const& src__, int ispn__, int i0__, int jspn__, int j0__){
+    assert(ispn__ == 0 || ispn__ == 1);
+    assert(jspn__ == 0 || jspn__ == 1);
+    std::cout << "=== WARNING at line " << __LINE__ << " of file " << __FILE__ << " ===" << std::endl;
+    std::cout << "    Copying Wavefunction with different type, possible lost of data precision" << std::endl;
+
+    int ngv = pw_coeffs(jspn__).num_rows_loc();
+    int nmt = has_mt() ? mt_coeffs(jspn__).num_rows_loc() : 0;
+
+    switch (pu__) {
+        case device_t::CPU: {
+            /* copy PW part */
+            std::copy(src__.pw_coeffs(ispn__).prime().at(memory_t::host, 0, i0__),
+                      src__.pw_coeffs(ispn__).prime().at(memory_t::host, 0, i0__) + ngv * n__,
+                      pw_coeffs(jspn__).prime().at(memory_t::host, 0, j0__));
+            /* copy MT part */
+            if (has_mt()) {
+                std::copy(src__.mt_coeffs(ispn__).prime().at(memory_t::host, 0, i0__),
+                          src__.mt_coeffs(ispn__).prime().at(memory_t::host, 0, i0__) + nmt * n__,
+                          mt_coeffs(jspn__).prime().at(memory_t::host, 0, j0__));
+            }
+            break;
+        }
+        case device_t::GPU: {
+            /* copy PW part */
+            acc::copy(pw_coeffs(jspn__).prime().at(memory_t::device, 0, j0__),
```

Comment:
On GPU, this is a memcpy operation. So this will not work for different types (T !=  F). A new GPU kernel for copying with type conversion might be required.